napari_btrack/sync.py

```diff
@@ -0,0 +1,117 @@
+"""
+This module contains functions for syncing widget values with TrackerConfig
+values.
+"""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from magicgui.widgets import Container
+
+    from napari_btrack.config import UnscaledTrackerConfig
+
+
+def update_config_from_widgets(
+    unscaled_config: UnscaledTrackerConfig,
+    container: Container,
+):
+    """Update an UnscaledTrackerConfig with the current widget values."""
+
+    sigmas = unscaled_config.sigmas
+    sigmas.P = container.P_sigma.value
+    sigmas.G = container.G_sigma.value
+    sigmas.R = container.R_sigma.value
+
+    config = unscaled_config.tracker_config
+    config.update_method = (
+        container.update_method_selector._widget._qwidget.currentIndex()
+    )
+    config.max_search_radius = container.max_search_radius.value
+
+    motion_model = config.motion_model
+    motion_model.accuracy = container.accuracy.value
+    motion_model.max_lost = container.max_lost.value
+
+    hypothesis_model = config.hypothesis_model
+    hypotheses = []
+    for hypothesis in [
+        "P_FP",
+        "P_init",
+        "P_term",
+        "P_link",
+        "P_branch",
+        "P_dead",
+        "P_merge",
+    ]:
+        if container[hypothesis].value:
+            hypotheses.append(hypothesis)
+    hypothesis_model.hypotheses = hypotheses
+
+    hypothesis_model.lambda_time = container.lambda_time.value
+    hypothesis_model.lambda_dist = container.lambda_dist.value
+    hypothesis_model.lambda_link = container.lambda_link.value
+    hypothesis_model.lambda_branch = container.lambda_branch.value
+
+    hypothesis_model.theta_dist = container.theta_dist.value
+    hypothesis_model.theta_time = container.theta_time.value
+    hypothesis_model.dist_thresh = container.dist_thresh.value
+    hypothesis_model.time_thresh = container.time_thresh.value
+    hypothesis_model.apop_thresh = container.apop_thresh.value
+    hypothesis_model.relax = container.relax.value
+
+    hypothesis_model.segmentation_miss_rate = container.segmentation_miss_rate.value
+
+    return unscaled_config
+
+
+def update_widgets_from_config(
+    unscaled_config: UnscaledTrackerConfig,
+    container: Container,
+):
+    """
+    Update the widgets in a container with the values in an
+    UnscaledTrackerConfig.
+    """
+
+    sigmas = unscaled_config.sigmas
+    container.P_sigma.value = sigmas.P
+    container.G_sigma.value = sigmas.G
+    container.R_sigma.value = sigmas.R
+
+    config = unscaled_config.tracker_config
+    container.update_method_selector.value = config.update_method.name
+    container.max_search_radius.value = config.max_search_radius
```

Comment:
I guess we could loop over the container widgets and use those names to get the attributes from the `TrackerConfig`. We could do this with `getattr`, but it would be nicer if `TrackerConfig` (in the main `btrack` package) had a `__get_item__` method defined so we could access its attributes like we would a dictionary. This would mean we could do:
```python
for widget in container:
    try:  # not all widgets have a corresponding attribute in `TrackerConfig`
        container[widget.name].value = config[widget.name]
    except KeyError as e:
        pass
```
Perhaps when we move this repo into the main `btrack` package we can at the same time add the necessary `__get_item__` methods to `TrackerConfig`, `MotionModel`, and `HypothesisModel` and simplify these update functions?

Reply:
I've replaced a lot of these lines with for-loops by defining some of the attributes in `napari_btrack.constants`. It's not so straightforward to do in one for-loop because there are lots of data structures we need to either update or take data from - `Sigmas`, `TrackerConfig`, `HypothesisModel`, `MotionModel`